Web Inspector: $event ReferenceError in Safari extension event listeners

dcrousso opened the pull request:
#### bec6d22a93435031499a6631639a841c3c438860
<pre>
Web Inspector: $event ReferenceError in Safari extension event listeners
<a href="https://bugs.webkit.org/show_bug.cgi?id=278919">https://bugs.webkit.org/show_bug.cgi?id=278919</a>

Reviewed by BJ Burg.

Grabbing the `GlobalObject` from the `ScriptExecutionContext` is not always going to be correct if the `EventListener` was created in a `DOMWrapperWorld`.

* Source/WebCore/inspector/agents/InspectorDOMDebuggerAgent.cpp:
(WebCore::globalObject): Added.
(WebCore::InspectorDOMDebuggerAgent::willHandleEvent):
(WebCore::InspectorDOMDebuggerAgent::didHandleEvent):

* LayoutTests/inspector/dom-debugger/command-line-api-event.html: Added.
* LayoutTests/inspector/dom-debugger/command-line-api-event-expected.txt: Added.

Canonical link: <a href="https://commits.webkit.org/283181@main">https://commits.webkit.org/283181@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/178d136a09f884fb28eb4e01d40b7705edd01782

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/65510 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/44881 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/18128 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/69536 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/59/builds/16119 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/67628 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/52682 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/16399 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/69536 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/59/builds/16119 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/68577 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/41458 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/56693 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/69536 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/38136 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/14071 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/14995 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/59970 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/14413 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/71241 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/9464 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/13869 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/71241 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/9496 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/56758 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/71241 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14433 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7819 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/1464 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/40691 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/41767 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/42950 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/41511 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

rcaliman-apple commented on Source/WebCore/inspector/agents/InspectorDOMDebuggerAgent.cpp:
```diff
@@ -250,12 +251,22 @@ Inspector::Protocol::ErrorStringOr<void> InspectorDOMDebuggerAgent::removeEventB
 
+static JSC::JSGlobalObject* globalObject(ScriptExecutionContext& scriptExecutionContext, EventListener& eventListener)
+{
+    if (auto* jsEventListener = dynamicDowncast<JSEventListener>(eventListener)) {
+        if (auto* isolatedWorld = jsEventListener->isolatedWorld())
+            return toJSDOMGlobalObject(scriptExecutionContext, *isolatedWorld);
+    }
```
Very neat! 👍 Thank you for providing a test for this!

burg commented on Source/WebCore/inspector/agents/InspectorDOMDebuggerAgent.cpp:
```diff
@@ -250,12 +251,22 @@ Inspector::Protocol::ErrorStringOr<void> InspectorDOMDebuggerAgent::removeEventB
     return makeUnexpected("Not supported"_s);
 }
 
+static JSC::JSGlobalObject* globalObject(ScriptExecutionContext& scriptExecutionContext, EventListener& eventListener)
```
Naming suggestion: globalObjectFor() or globalObjectFromListenerWorld()

dcrousso committed:
Web Inspector: $event ReferenceError in Safari extension event listeners
https://bugs.webkit.org/show_bug.cgi?id=278919

Reviewed by BJ Burg.

Grabbing the `GlobalObject` from the `ScriptExecutionContext` is not always going to be correct if the `EventListener` was created in a `DOMWrapperWorld`.

* Source/WebCore/inspector/agents/InspectorDOMDebuggerAgent.cpp:
(WebCore::globalObject): Added.
(WebCore::InspectorDOMDebuggerAgent::willHandleEvent):
(WebCore::InspectorDOMDebuggerAgent::didHandleEvent):

* LayoutTests/inspector/dom-debugger/command-line-api-event.html: Added.
* LayoutTests/inspector/dom-debugger/command-line-api-event-expected.txt: Added.

Canonical link: https://commits.webkit.org/283181@main
